chore: ゲートを新設・変更したらミューテーションで発火を確認する、を .claude/rules へ規約化する

## 概要

**ゲートを新設・変更したら、その場でミューテーションして発火することを確認する**——という規約が
`.claude/rules/` に無い。結果として「入れたつもりのゲートが一度も発火していない」が繰り返し起きている。

現状 `.claude/rules/architecture.md` には「ArchUnit で Kotlin の呼び出しを縛るときの空振り」
（`noClasses().should()` はマッチ 0 件ならパスする）という**手法固有**の警告があるだけで、
detekt カスタムルール・Claude hook・CI ジョブなど他のゲートには及んでいない。

## 実績（同じ質の問題が 3 件）

| Issue | ゲート | 空振りの中身 | 気づいた経緯 |
| --- | --- | --- | --- |
| #519 → #780 | Claude Stop hook の並行実行ガード | 判定文字列 `org.gradle.wrapper.GradleWrapperMain` がどのプロセスにも当たらず、**導入以降ただの一度も発火していなかった** | 防ぐはずの競合が #748 / PR #778 で実際に起きた（偽の BUILD FAILED × 3・約 15 分の浪費） |
| #754 | ArchUnit `domainIsFrameworkFree` | 禁止列挙が Jackson 2（`com.fasterxml.jackson..`）のみで、実際に使う Jackson 3（`tools.jackson..`）を取りこぼす | #750 のレビューが「ArchUnit が禁じている」という根拠を検証しにいって発覚 |
| #737 | jMolecules アノテーション | 付け忘れが静かに通る（そもそもゲートが無い） | — |

共通するのは **「ゲートが緑であること自体が根拠として使われる」** 点。このリポジトリは「規約は機械強制する」
前提で運用しているため、空振りしているゲートは *守れているように見えて実は何もしていない* 状態になり、
後続の判断まで汚染する。#780 では、偽の BUILD FAILED を見たエージェントが自分の変更を疑って
原因を探しにいくという二次被害まで出た。

## やること

- `.claude/rules/` に、ゲート全般（ArchUnit / detekt カスタムルール / Claude hook / CI ジョブ / lefthook）を
  対象とした「非空振りの確認」規約を書く。骨子:
  - ゲートを新設・変更したら、**わざと違反する状態を作って落ちることを見る**まで完了としない
  - 確認したことを PR 本文に実測（出力の抜粋）付きで残す
  - 「緑だった」は非空振りの根拠にならない
- 置き場所を決める。候補は `.claude/rules/testing.md`（ゲート＝検証手段という括り）か
  `.claude/rules/architecture.md`（既存の ArchUnit 空振り警告と同居）。既存記述との重複を避けつつ、
  ArchUnit 固有だった警告を一般化して吸収する形が望ましい
- 常時ロードの footprint を増やさないこと（CLAUDE.md 本体には置かず、path-scope 付きの rules に置く）

## 関連

- #780 / PR #781 hook のガードが構造上発火しなかった件（今回の起点）。ミューテーション確認が実際に有効だった実例でもある
- #519 / PR #524 そのガードを「対策済み」としてクローズした作業
- #754 ArchUnit が Jackson 3 を取りこぼす
- #737 jMolecules の付け忘れが静かに通る
- #765 #687 で判明した規約知見を `.claude/rules` へ書き戻す（別テーマの同種作業）

--- a/.claude/rules/architecture.md
+++ b/.claude/rules/architecture.md
@@ -180,4 +180,4 @@ class Command<T>(val payload: T, val issuedAt: Instant)
 - **inline value class によるメソッド名マングリング**: `@JvmInline value class`（ID 値クラス等）を**引数に取るメソッドも、戻り値に持つメソッドも**、JVM 名がハッシュ付きにマングルされる（`of` → `of-Havw-KM`、`Result` を返す `invoke` → `invoke-Zyo9ksc`）。`haveName("invoke")` や `target.name == "of"` は完全一致せず空振りする。ベース名（`name.substringBefore('-')`）で突合するか `haveNameStartingWith` で前方一致させ、過剰マッチはパラメータ型条件との AND で絞る。
 - **companion object メソッドの owner**: `Foo.of(...)` の呼び出しターゲット owner は `Foo` ではなく `Foo$Companion`。`call.target.owner.enclosingClass.orElse(owner)` で囲みクラスを辿って突合する。
 
-**新しく書いた呼び出し制約ルールは、必ずミューテーションで非空振りを確認する**。許可パッケージ定数を一時的に bogus 値へ変えるか、規約を守っているコードのアノテーションを外し、テストが FAIL することを確かめてから戻す。パスしたまま BUILD SUCCESSFUL なら述語が空振りしている。恒久的には違反サンプル fixture ＋ `assertNotSatisfied` 系の回帰テストを別途置く（先例: `AggregateNotDataClassRuleTest` / `DtoDomainEnumRuleTest`）。
+**新しく書いた呼び出し制約ルールは、必ずミューテーションで非空振りを確認する**。パスしたまま BUILD SUCCESSFUL なら述語が空振りしている。壊し方・恒久的な回帰テストの置き方・他のゲート種別（detekt / Claude hook / CI ジョブ / lefthook）にも共通する一般則は `.claude/rules/gates.md` にある。

--- a/.claude/rules/gates.md
+++ b/.claude/rules/gates.md
@@ -1,0 +1,47 @@
+---
+paths:
+  - "src/test/**"
+  - "detekt-rules/src/**"
+  - ".github/workflows/**"
+  - "lefthook.yml"
+  - ".claude/hooks/**"
+  - ".claude/settings.json"
+  - ".claude/rules/gates.md"
+---
+
+# ゲートの非空振り規約
+
+このリポジトリは「規約は機械強制する」前提で運用している。したがって**空振りしているゲートは、守れているように見えて実は何もしていない**状態になり、「ゲートが緑である」ことを根拠に使う後続の判断まで汚染する（#780 では、空振りしていたガードが防ぐはずの競合が実際に起き、偽の BUILD FAILED を見たエージェントが自分の変更を疑って原因を探しにいく二次被害まで出た）。
+
+対象は ArchUnit / detekt カスタムルール / Claude hook / CI ジョブ / lefthook のすべて。
+
+## 原則
+
+- **ゲートを新設・変更したら、わざと違反する状態を作って落ちることを見るまで完了としない**。緑のまま通ったなら、それは「違反が無い」かもしれないし「検出が空振りしている」かもしれない。両者は出力で区別できない
+- **確認したことは PR 本文に実測（FAIL したときの出力の抜粋）付きで残す**。「確認した」という記述は後から検証できないが、出力は検証できる
+- **「緑だった」は非空振りの根拠にならない**。ゲートを導入した PR が緑であることは、そのゲートが動いた証拠ではない（#780 のガードは導入以降ただの一度も発火していなかった）
+
+## 種別ごとのミューテーションのやり方
+
+「何を壊せば FAIL するはずか」を先に決めてから壊す。想定どおりに落ちなければ、落ちない理由が分かるまでゲートは未完成とみなす。このとき**そもそも違反が成立しているか**を先に疑う。壊し方を間違えると「ゲートが空振りしている」ではなく「違反になっていないので落ちない」を見ることになり、正しいゲートを壊れていると誤診する（実測: 下の Claude hook の例を `.md` の行末空白で試すと exit 0 になる。`.editorconfig` が `[*.md] trim_trailing_whitespace = false` としているため違反ではない。`.kt` で試せば exit 2 で発火する）。
+
+- **ArchUnit**: 許可パッケージ定数を一時的に bogus 値へ変えるか、規約を守っているコードのアノテーションを外し、テストが FAIL することを確かめてから戻す。`noClasses().should()` はマッチが 0 件なら違反 0 件＝パスするため、述語のバグは無言の偽陰性になる。Kotlin 固有の空振り原因（inline value class のメソッド名マングリング・companion object の owner）は `.claude/rules/architecture.md` の「ArchUnit で Kotlin の呼び出しを縛るときの空振り」にある
+- **detekt カスタムルール**: `detekt-test` の `rule.lint(...)` に違反スニペットを食わせ、`findings` が期待どおり出ることと、適合コードでは出ないことの両方を見る（先例: `ActorScopedUseCaseTest` / `WorldScopedPortSignatureTest`）
+- **Claude hook**: hook は stdin の JSON で駆動するので、想定するペイロードを流し込んで単体実行し、exit code と stderr を見る。
+
+  ```bash
+  echo '{"tool_input":{"file_path":"path/to/violating.kt"}}' | .claude/hooks/post-edit-editorconfig.sh
+  echo $?   # 違反を検出したなら 2
+  ```
+
+  実行中のプロセスやリポジトリの状態を判定に使う hook は、**その判定文字列が実際に何にマッチするか**まで確かめる（#780 のガードは `org.gradle.wrapper.GradleWrapperMain` がどのプロセスにも当たらず、構文としては正しいまま一度も発火しなかった）
+- **CI ジョブ**: 検査の本体はワークフローのインラインではなく `scripts/*.sh` へ切り出す。違反状態を作ってローカルで直接実行するだけで発火を確認できる（先例: `scripts/check-adr-numbering.sh` と `.github/workflows/adr-check.yml`）。加えて **`paths` フィルタがその変更でジョブを起動するか**を別途確かめる。フィルタが外れているとゲート本体が正しくてもジョブごと走らない
+- **lefthook**: 違反ファイルを stage して `lefthook run pre-commit`（push 側は `lefthook run pre-push`）を実行し、当該コマンドが skip されずに落ちることを見る。**glob が効かず、対象ファイルだけのコミットで素通りする**穴が繰り返し出ている（#800 / #804）。glob を書いたら「そのファイル 1 本だけを stage したとき」に走るかを必ず確かめる
+
+## 一度きりのミューテーションで終わらせない
+
+一時的に壊して戻す確認は、そのとき見た空振りしか防げない。違反サンプルを恒久的な fixture として置ける種別（ArchUnit / detekt カスタムルール）では、FAIL 側を検証する回帰テスト（ArchUnit なら違反サンプル fixture ＋ `assertNotSatisfied` 系、detekt なら違反スニペットの `lint`）を併せて置く（先例: `AggregateNotDataClassRuleTest` / `DtoDomainEnumRuleTest` / `ControllerPackageLayoutRuleTest`）。
+
+## この規約自体について
+
+この規約はレビュー担保であり、機械強制していない（「ミューテーションを実施したか」を検査するゲートは無い）。

